### PR TITLE
[eclipse/xtext-extras#695] fix constructor replacement in aa

### PR DESCRIPTION
**Pull request by cdietrich**
[WIP][eclipse/xtext-extras#695] fix constructor replacement in aa
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/declaration/JvmTypeDeclarationImpl.java [hunk 1]
```diff
@@ -78,6 +79,7 @@ public MutableConstructorDeclaration addConstructor(final Procedure1<MutableCons
     };
     final JvmConstructor constructor = IterableExtensions.<JvmConstructor>findFirst(Iterables.<JvmConstructor>filter(this.getDelegate().getMembers(), JvmConstructor.class), _function);
     if ((constructor != null)) {
+      final EObject src = this.getCompilationUnit().getJvmModelAssociations().getPrimarySourceElement(constructor.getDeclaringType());
```
Comment (cdietrich) [inline org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/declaration/JvmTypeDeclarationImpl.java:82]: am not sure about removeAllAssociation vs removeAssociation

Comment (cdietrich) [reply org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/declaration/JvmTypeDeclarationImpl.java:82]: removeAllAssociation does not work in domain model example cause 1:n

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend [hunk 2]
```diff
@@ -236,6 +236,9 @@ abstract class JvmTypeDeclarationImpl<T extends JvmDeclaredType> extends JvmMemb
 		// remove default constructor
 		val constructor = delegate.members.filter(JvmConstructor).findFirst[compilationUnit.typeExtensions.isSingleSyntheticDefaultConstructor(it)]
 		if (constructor !== null) {
+			// val src = compilationUnit.jvmModelAssociations.getPrimarySourceElement(constructor.declaringType)
```
Comment (cdietrich) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:239]: am not sure which one is correct or better.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend [hunk 3]
```diff
@@ -236,6 +236,9 @@ abstract class JvmTypeDeclarationImpl<T extends JvmDeclaredType> extends JvmMemb
...
+			// compilationUnit.jvmModelAssociator.removeAssociation(src, constructor)
+			compilationUnit.jvmModelAssociator.removeAllAssociation(constructor)
```
Comment (cdietrich) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:241]: removeAll seems to correctly clean the orphans, the remove only with the fix in extras

Comment (szarnekow) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:241]: Is there a chance to use `org.eclipse.xtend.core.macro.declaration.JvmElementImpl.remove()` instead?

Comment (szarnekow) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:241]: Probably not worth it. removeAll appears to be the right thing to do.
We could introduce a package visible `compilationUnit.removeJvmElement(EObject)` and call that one from `JvmElementImpl.remove` and from here, though.

**Review by szarnekow — APPROVED** [review]

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend [hunk 4]
```diff
@@ -236,7 +236,8 @@ abstract class JvmTypeDeclarationImpl<T extends JvmDeclaredType> extends JvmMemb
 		// remove default constructor
 		val constructor = delegate.members.filter(JvmConstructor).findFirst[compilationUnit.typeExtensions.isSingleSyntheticDefaultConstructor(it)]
 		if (constructor !== null) {
-			EcoreUtil.remove(constructor)
+			val mutableConstructorDeclaration = compilationUnit.toMemberDeclaration(constructor) as MutableConstructorDeclaration
```
Comment (cdietrich) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:239]: @szarnekow better this way?

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend]: yes this worked

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend (hunk 4 above)
Comment (szarnekow) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/declaration/JvmBasedDeclarations.xtend:239]: I do think this makes sense, yes.

**Review by szarnekow — APPROVED** [review]

**Pull request merged**